File: Libraries/python/beagle_helper/src/beagle_helper/devices/seven_segment.py
from pathlib import Path


class SevenSegment:
    def __init__(self, p: Path) -> None:
        if not p.is_dir():
            raise ValueError("Seven segment path does not exist")

        self.scroll_step_path = p.joinpath("scroll_step_ms")
        self.message_path = p.joinpath("message")

    def set_message(self, msg: str):
        with open(self.message_path, 'w') as f:
            f.write(msg)

    def set_step(self, step_ms: int):
        with open(self.scroll_step_path, 'w') as f:
            f.write(str(step_ms))

File: Libraries/rust/beagle_helper/src/devices/seven_segment.rs
use std::{
    fs::{File, OpenOptions},
    io::{Seek, Write},
    path::Path,
    time::Duration,
};

pub struct SevenSegment {
    scroll_step_file: File,
    message_file: File,
}

impl SevenSegment {
    pub fn new(path: &str) -> std::io::Result<Self> {
        let base_dir = Path::new(path);

        if !base_dir.is_dir() {
            return Err(std::io::Error::new(
                std::io::ErrorKind::NotADirectory,
                "Base path should be directory",
            ));
        }

        let scroll_step_file = OpenOptions::new()
            .read(true)
            .write(true)
            .open(base_dir.join("scroll_step_ms"))?;
        let message_file = OpenOptions::new()
            .read(true)
            .write(true)
            .open(base_dir.join("message"))?;

        Ok(Self {
            scroll_step_file,
            message_file,
        })
    }

    pub fn set_message(&mut self, msg: &str) -> std::io::Result<()> {
        let _ = self.message_file.seek(std::io::SeekFrom::Start(0));
        self.message_file.write_all(msg.as_bytes())?;
        Ok(())
    }

    pub fn set_step(&mut self, step: Duration) -> std::io::Result<()> {
        let _ = self.scroll_step_file.seek(std::io::SeekFrom::Start(0));
        self.scroll_step_file
            .write_all(step.as_millis().to_string().as_bytes())?;
        Ok(())
    }
}
